tests/byoc/byoc.go

```diff
@@ -0,0 +1,260 @@
+		ApplicationService: ApplicationService{
+			GithubRepository: QuarkusDevfileSource,
+			ApplicationName:  applicationName,
+		},
+		Byoc: Byoc{
+			ClusterType:     appservice.ConfigurationClusterType_Kubernetes,
+			TargetNamespace: utils.GetGeneratedNamespace("byoc-k8s-target"),
+		},
+	},
+	{
+		Name: "Deploy RHTAP sample application into a Openshift cluster provided by user",
+		ApplicationService: ApplicationService{
+			GithubRepository: QuarkusDevfileSource,
+			ApplicationName:  applicationName,
+		},
+		Byoc: Byoc{
+			ClusterType:     appservice.ConfigurationClusterType_OpenShift,
+			TargetNamespace: utils.GetGeneratedNamespace("byoc-ocp-target"),
+		},
+	},
+}
+
+var _ = framework.E2ESuiteDescribe(Label("byoc"), Ordered, func() {
+	rootPath, err := os.Getwd()
+	Expect(err).NotTo(HaveOccurred())
+
+	var vc vcluster.Vcluster
+	var ephemeralClusterClient *kubernetes.Clientset
+	var fw *framework.Framework
+	var byocKubeconfig, byocAPIServerURL string
+
+	// Initialize the application struct
+	application := &appservice.Application{}
+	cdq := &appservice.ComponentDetectionQuery{}
+	compDetected := appservice.ComponentDetectionDescription{}
+	componentObj := &appservice.Component{}
+
+	for _, suite := range byocTestScenario {
+		suite := suite
+
+		Describe(suite.Name, func() {
+			BeforeAll(func() {
+				fw, err = framework.NewFramework(utils.GetGeneratedNamespace(fmt.Sprintf("byoc-%s", strings.ToLower(string(suite.Byoc.ClusterType)))))
+				Expect(err).NotTo(HaveOccurred())
+
+				if suite.Byoc.ClusterType == appservice.ConfigurationClusterType_Kubernetes {
+					vc = vcluster.NewVclusterController(fmt.Sprintf("%s/tmp", rootPath), fw.AsKubeAdmin.CommonController.CustomClient)
+				}
+			})
+
+			// Remove all resources created by the tests
+			AfterAll(func() {
+				if !CurrentSpecReport().Failed() {
+					Expect(fw.AsKubeDeveloper.HasController.DeleteAllComponentsInASpecificNamespace(fw.UserNamespace, 30*time.Second)).To(Succeed())
+					Expect(fw.AsKubeAdmin.HasController.DeleteAllApplicationsInASpecificNamespace(fw.UserNamespace, 30*time.Second)).To(Succeed())
+					Expect(fw.AsKubeAdmin.HasController.DeleteAllSnapshotEnvBindingsInASpecificNamespace(fw.UserNamespace, 30*time.Second)).To(Succeed())
+					Expect(fw.AsKubeAdmin.ReleaseController.DeleteAllSnapshotsInASpecificNamespace(fw.UserNamespace, 30*time.Second)).To(Succeed())
+					Expect(fw.AsKubeAdmin.GitOpsController.DeleteAllEnvironmentsInASpecificNamespace(fw.UserNamespace, 30*time.Second)).To(Succeed())
+					Expect(fw.AsKubeAdmin.TektonController.DeleteAllPipelineRunsInASpecificNamespace(fw.UserNamespace)).To(Succeed())
+					Expect(fw.AsKubeAdmin.GitOpsController.DeleteAllGitOpsDeploymentInASpecificNamespace(fw.UserNamespace, 30*time.Second)).To(Succeed())
+					Expect(fw.SandboxController.DeleteUserSignup(fw.UserName)).NotTo(BeFalse())
+				}
+			})
+
+			It("initializes byoc cluster connection", func() {
```

Comment:
Is this more of a precondition? 

Same comment about the Cluster class and abstract factory pattern mentioned above.